gateway-ha/src/main/java/io/trino/gateway/ha/handler/RoutingTargetHandler.java

```diff
@@ -66,13 +66,27 @@ public RoutingTargetHandler(
         cookiesEnabled = GatewayCookieConfigurationPropertiesProvider.getInstance().isEnabled();
     }
 
-    public String getRoutingDestination(HttpServletRequest request)
+    public RoutingDestination getRoutingDestination(HttpServletRequest request)
     {
-        Optional<String> previousBackend = getPreviousBackend(request);
-        String clusterHost = previousBackend.orElseGet(() -> getBackendFromRoutingGroup(request));
-        logRewrite(clusterHost, request);
+        Optional<String> queryId = extractQueryIdIfPresent(request, statementPaths, requestAnalyserClientsUseV2Format, requestAnalyserMaxBodySize);
+        Optional<String> previousCluster = getPreviousCluster(queryId, request);
+        RoutingDestination routingDestination = previousCluster.map(cluster -> {
+            String routingGroup = queryId.map(routingManager::findRoutingGroupForQueryId)
+                    .orElse("adhoc");
```

Comment:
@willmostly in the case where `previousCluster` is found but `queryId` is not found (i.e., cookie-based routing), how should we handle the value for routing group?